examples/vae_text/vae_train.py

```diff
@@ -0,0 +1,463 @@
+            output_embed = torch.cat([output_embed, latent_z], 2)
+
+            train_helper = self.decoder.create_helper(
+                decoding_strategy='train_greedy')
+            outputs, _, _ = self.decoder(
+                initial_state=dcdr_states,
+                helper=train_helper,
+                inputs=output_embed,
+                sequence_length=data_batch["length"] - 1)
+        else:
+            outputs = self.decoder(
+                inputs=output_embed,
+                memory=dcdr_states,
+                memory_sequence_length=torch.ones(dcdr_states.size(0)))
+
+        logits = outputs.logits
+
+        seq_lengths = data_batch["length"] - 1
+        # Losses & train ops
+        rc_loss = tx.losses.sequence_sparse_softmax_cross_entropy(
+            labels=text_ids[:, 1:],
+            logits=logits,
+            sequence_length=data_batch["length"] - 1)
+
+        nll = rc_loss + kl_weight * kl_loss
+
+        fetches = {
+            "nll": nll,
+            "kl_loss": kl_loss.detach(),
+            "rc_loss": rc_loss.detach(),
+            "lengths": seq_lengths.detach()
+        }
+
+        return fetches
+
+
+def main():
+    # Data
+    train_data = tx.data.MonoTextData(config.train_data_hparams,
+                                      device=device)
+    val_data = tx.data.MonoTextData(config.val_data_hparams,
+                                    device=device)
+    test_data = tx.data.MonoTextData(config.test_data_hparams,
+                                     device=device)
+
+    iterator = tx.data.TrainTestDataIterator(train=train_data,
+                                             val=val_data,
+                                             test=test_data)
+
+    opt_vars = {
+        'learning_rate': config.lr_decay_hparams["init_lr"],
+        'best_valid_nll': 1e100,
+        'steps_not_improved': 0,
+        'kl_weight': config.kl_anneal_hparams["start"]
+    }
+
+    decay_cnt = 0
+    max_decay = config.lr_decay_hparams["max_decay"]
+    decay_factor = config.lr_decay_hparams["decay_factor"]
+    decay_ts = config.lr_decay_hparams["threshold"]
+
+    save_dir = f"./models/{config.dataset}"
+
+    if not os.path.exists(save_dir):
+        os.makedirs(save_dir)
+
+    suffix = f"{config.dataset}_{config.decoder_type}Decoder.ckpt"
+
+    save_path = os.path.join(save_dir, suffix)
+
+    # KL term annealing rate
+    anneal_r = 1.0 / (config.kl_anneal_hparams["warm_up"] *
+        (train_data.__len__() / config.batch_size))
+
+    model = Vae(train_data)
+    model.to(device)
+
+    optimizer = tx.core.get_optimizer(params=model.parameters(),
+                                    hparams=config.opt_hparams)
+
+    def _run_epoch(epoch, mode, display=10):
+        if mode == 'train':
+            iterator.switch_to_train_data()
+        elif mode == 'valid':
+            iterator.switch_to_val_data()
+        elif mode == 'test':
+            iterator.switch_to_test_data()
```

Comment:
Let's just use `DataIterator` and call `iterator.switch_to_dataset(mode)` here.

Reply:
`DataIterator` is implemented.